src/model64.c

```diff
@@ -740,38 +688,185 @@ static void quat_lerp(float *out, float *in1, float *in2, float time)
+    uint32_t num_tracks = curr_anim->num_tracks;
+    uint32_t frame_buf_size = (num_tracks*4)*sizeof(model64_keyframe_t);
+    if(model->data->anim_data_handle) {
+        uint32_t rom_addr = (uint32_t)model->data->anim_data_handle;
+        rom_addr += (uint32_t)curr_anim->keyframes;
+        data_cache_hit_writeback_invalidate(anim_state->frames, frame_buf_size);
+        dma_read(anim_state->frames, rom_addr, frame_buf_size);
+    } else {
+        memcpy(anim_state->frames, curr_anim->keyframes, frame_buf_size);
+    }
+    for(uint32_t i=0; i<num_tracks; i++) {
+        anim_state->buf_idx[i] = 0;
+    }
+    anim_state->frame_idx = num_tracks*4;
+    anim_state->prev_waiting_frame = false;
+    if(anim_state->frame_idx >= curr_anim->num_keyframes) {
+        memcpy(anim_state->waiting_frame, &anim_state->frames[curr_anim->num_keyframes-1], frame_buf_size);
+        anim_state->frame_idx = curr_anim->num_keyframes;
+    } else {
+        read_waiting_frame(model, anim_slot);
+    }
+    
+}
+
+static void fetch_needed_keyframes(model64_t *model, model64_anim_slot_t anim_slot)
+{
+    anim_state_t *anim_state = model->active_anims[anim_slot];
+    model64_anim_t *curr_anim = &model->data->anims[anim_state->index];
+    while(anim_state->waiting_frame->time_req < anim_state->time) {
+        if(anim_state->frame_idx >= curr_anim->num_keyframes) {
+            return;
+        }
+        read_waiting_frame(model, anim_slot);
+    }
+}
+
+static void calc_anim_pose(model64_t *model, model64_anim_slot_t anim_slot)
+{
+    anim_state_t *anim_state = model->active_anims[anim_slot];
+    model64_anim_t *curr_anim = &model->data->anims[anim_state->index];
+    float decoded_values[4][4];
+    for(uint32_t i=0; i<curr_anim->num_tracks; i++) {
+        uint32_t component = curr_anim->tracks[i] >> 14;
+        uint16_t node = curr_anim->tracks[i] & 0x3FFF;
+        for(uint32_t j=0; j<4; j++) {
+            int index = (i*4)+((j+anim_state->buf_idx[i]) % 4);
+            model64_keyframe_t *frame = &anim_state->frames[index];
+            switch(component) {
+                case ANIM_COMPONENT_POS:
+                    decode_normalized_u16(frame->data, &decoded_values[j][0], 3, curr_anim->pos_min, curr_anim->pos_max);
+                    decoded_values[j][3] = 0.0f;
+                    break;
+                    
+                case ANIM_COMPONENT_ROT:
+                    decode_quaternion(frame->data, &decoded_values[j][0]);
+                    break;
+                    
+                case ANIM_COMPONENT_SCALE:
+                    decode_normalized_u16(frame->data, &decoded_values[j][0], 3, curr_anim->scale_min, curr_anim->scale_max);
+                    decoded_values[j][3] = 0.0f;
+                    break;
+                  
+                default:
+                    decoded_values[j][0] = decoded_values[j][1] = decoded_values[j][2] = decoded_values[j][3] = 0.0f;
+                    break;
+            }
+        }
```

Comment:
I think it's possible to avoid doing most of the above work at each `calc_anim_pose` call, and just doing it when a new keyframe arrives. `decoded_values` can be stored somewhere in `anim_state` so that `calc_anim_pose` can just perform an interpolation.